File: Ogury/SASOguryInterstitialAdapter.m
//
//  SASOguryInterstitialAdapter.m
//  AdViewer
//
//  Created by Loïc GIRON DIT METAZ on 29/09/2020.
//  Copyright © 2020 Smart AdServer. All rights reserved.
//

#import "SASOguryInterstitialAdapter.h"

NS_ASSUME_NONNULL_BEGIN

@implementation SASOguryInterstitialAdapter

- (instancetype)initWithDelegate:(id<SASMediationInterstitialAdapterDelegate>)delegate {
    if (self = [super init]) {
        self.delegate = delegate;
    }
    return self;
}

- (void)requestInterstitialWithServerParameterString:(NSString *)serverParameterString clientParameters:(NSDictionary *)clientParameters {
    // Ogury configuration
    NSError *error = nil;
    if (![self configureOgurySDKWithServerParameterString:serverParameterString andClientParameters:clientParameters error:&error]) {
        // Configuration can fail if the serverParameterString is invalid
        [self.delegate mediationInterstitialAdapter:self didFailToLoadWithError:error noFill:NO];
        return;
    }
    
    // Interstitial instantiation and loading
    self.interstitial = [[OguryAdsInterstitial alloc] initWithAdUnitID:self.adUnitId];
    self.interstitial.interstitialDelegate = self;
    
    [self.interstitial load];
}

- (void)showInterstitialFromViewController:(UIViewController *)viewController {
    [self.interstitial showInViewController:viewController];
}

- (BOOL)isInterstitialReady {
    if (self.interstitial != nil) {
        return self.interstitial.isLoaded;
    } else {
        return NO;
    }
}

#pragma mark - Ogury interstitial delegate

- (void)oguryAdsInterstitialAdAvailable { }

- (void)oguryAdsInterstitialAdNotAvailable {
    NSError *error = [NSError errorWithDomain:SASOguryAdapterErrorDomain code:SASOguryAdapterErrorCodeAdNotAvailable userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Ogury Interstitial - Ad not available"] }];
    [self.delegate mediationInterstitialAdapter:self didFailToLoadWithError:error noFill:YES];
}

- (void)oguryAdsInterstitialAdLoaded {
    [self.delegate mediationInterstitialAdapterDidLoad:self];
}

- (void)oguryAdsInterstitialAdNotLoaded {
    NSError *error = [NSError errorWithDomain:SASOguryAdapterErrorDomain code:SASOguryAdapterErrorCodeAdNotLoaded userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Ogury Interstitial - Ad not loaded"] }];
    [self.delegate mediationInterstitialAdapter:self didFailToLoadWithError:error noFill:NO];
}

- (void)oguryAdsInterstitialAdDisplayed {
    [self.delegate mediationInterstitialAdapterDidShow:self];
}

- (void)oguryAdsInterstitialAdClosed {
    [self.delegate mediationInterstitialAdapterDidClose:self];
}

- (void)oguryAdsInterstitialAdError:(OguryAdsErrorType)errorType {
    NSError *error = [NSError errorWithDomain:SASOguryAdapterErrorDomain code:SASOguryAdapterErrorCodeAdError userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Ogury Interstitial - Ad error with type: %ld", errorType] }];
    [self.delegate mediationInterstitialAdapter:self didFailToLoadWithError:error noFill:NO];
}

- (void)oguryAdsInterstitialAdClicked {
    [self.delegate mediationInterstitialAdapterDidReceiveAdClickedEvent:self];
}

@end

NS_ASSUME_NONNULL_END

File: Ogury/SASOguryInterstitialAdapter.swift
//
//  SASOguryInterstitialAdapter.swift
//  SmartAdServer
//
//  Created by Loïc GIRON DIT METAZ on 21/07/2021.
//  Copyright © 2021 Smart AdServer. All rights reserved.
//

import UIKit
import SASDisplayKit
import OguryAds

/**
 Ogury interstitial adapter class for Smart SDK mediation.
 
 Mediation adapter classes can be used to display an ad using a third party SDK directly from
 an insertion handled by Smart.
 
 To use an adapter class, you simply have to add them to your Xcode project and they will
 be automatically instantiated by the Smart SDK if needed.
 */
@objc(SASOguryInterstitialAdapter)
class SASOguryInterstitialAdapter: SASOguryBaseAdapter, SASMediationInterstitialAdapter {
    
    /// A delegate that this adapter must call to provide information about the ad loading status or events to the Smart Display SDK.
    private weak var delegate: SASMediationInterstitialAdapterDelegate? = nil
    
    /// The currently loaded Ogury interstitial if any, nil otherwise.
    private var interstitial: OguryInterstitialAd? = nil
    
    required init(delegate: SASMediationInterstitialAdapterDelegate) {
        self.delegate = delegate
        super.init()
    }
    
    func requestInterstitial(withServerParameterString serverParameterString: String, clientParameters: [AnyHashable : Any]) {
        // Ogury configuration
        configureOgurySDK(serverParameterString: serverParameterString, clientParameters: clientParameters) { [self] error in
            if let error = error {
                // Configuration can fail if the serverParameterString is invalid or if the Ogury SDK does not initialize properly
                delegate?.mediationInterstitialAdapter(self, didFailToLoadWithError: error, noFill: false)
            } else {
                // Interstitial instantiation and loading
                interstitial = OguryInterstitialAd(adUnitId: adUnitId!)
                interstitial?.delegate = self
                
                interstitial?.load()
            }
        }
    }
    
    func showInterstitial(from viewController: UIViewController) {
        interstitial?.show(in: viewController)
    }
    
    func isInterstitialReady() -> Bool {
        if let interstitial = interstitial {
            return interstitial.isLoaded()
        } else {
            return false
        }
    }
    
}

/**
 Ogury delegate implementation.
 */
extension SASOguryInterstitialAdapter: OguryInterstitialAdDelegate {
    
    func didLoad(_ interstitial: OguryInterstitialAd) {
        delegate?.mediationInterstitialAdapterDidLoad(self)
    }
    
    func didDisplay(_ interstitial: OguryInterstitialAd) {
        delegate?.mediationInterstitialAdapterDidShow(self)
    }
    
    func didClick(_ interstitial: OguryInterstitialAd) {
        delegate?.mediationInterstitialAdapterDidReceiveAdClickedEvent(self)
    }
    
    func didClose(_ interstitial: OguryInterstitialAd) {
        delegate?.mediationInterstitialAdapterDidClose(self)
    }
    
    func didFailOguryInterstitialAdWithError(_ error: OguryError, for interstitial: OguryInterstitialAd) {
        let error = NSError(
            domain: ErrorConstants.errorDomain,
            code: ErrorConstants.errorCodeAdError,
            userInfo: [NSLocalizedDescriptionKey: "Ogury Interstitial - Ad failed with error: \(error)"]
        )
        delegate?.mediationInterstitialAdapter(self, didFailToLoadWithError: error, noFill: (error.code == ErrorConstants.oguryNoAdErrorCode))
    }
    
    func didTriggerImpressionOguryInterstitialAd(_ interstitial: OguryInterstitialAd) { }
    
}
